### DISCUSSION TITLE
How to use other Functions for interacting with the game?

### DISCUSSION THREAD

**Discussion by Jajo006** [PrismarineJS/bedrock-protocol#306] [Q&A]
Hey Guys,
I had to edit my question because it was unclear and I figured out a lot more about what I wan't to do.
My original question was how to send other packets as a client to the server for eg. player movement.

What I figured out by now is that you can send all client packets from [https://minecraft-data.prismarine.js.org/?v=bedrock_1.19.30&d=protocol](url) to the server. The problem is that I can't get that to work. I've tried to make the client rotate with `move_player` with this packet:
`  client.queue('move_player', {
    runtime_entity_id: client.startGameData.runtime_entity_id,
    position: {x:0,y:0,z:0},
    pitch: 20,
    yaw: 20,
    head_yaw: 20,
    mode: 3,
    on_ground: true,
    ridden_runtime_id: 0,
    teleport: undefined,
    tick: client.tick
  })`
Did I make a mistake? 

Thanks for any help :)

**Comment by frizib** [comment]
Hi. Did you manage to do it?

**Reply by Jajo006** [reply]
Yes, I managed to move and rotate the player on my dedicated bedrock server.
I asked around on their discord server and they suggested that I use the player_input packet. I could send you an example but I‘m not home right now.

Edit: they also suggested that to find out how to send the packets correctly you can use a relay example like in the readme to receive „real“ packets send by a real game. Again, I can help you in Detail when I get home.
Also sorry for the bad spelling, german autocorrect is turning every sentence 180.

**Reply by frizib** [reply]
Are there any examples for mitm proxy?

**Reply by Jajo006** [reply]
Hi, I could send you an example for the proxy but I don‘t know if I still have one .
The packet I used for movement was the `player_auth_input` packet.
I wrote this example a while ago but as far as I know, it still works.
It basically moves in a direction dependent on a chat input like f for forward.
Let me know if it worked for you:
```const bedrock = require('bedrock-protocol')
var Vec3 = require('vec3').Vec3;
const client = bedrock.createClient({
  host: 'YOUR IP',   // optional
  port: 19132,         // optional, default 19132
  offline: false       // optional, default false. if true, do not login with Xbox Live. You will not be asked to sign-in if set to true.
})


let pos = {}
let offsetx = 0
let offsety = 0

client.on('respawn', (packet) => {
  console.log(packet.position);
  pos = packet.position;
})

//client.on('add_entity', (packet) => {
//  console.log(packet.entity_type)
//})

client.on('text', (packet) => {
  if(packet.message=='b'){
    offsetx = -1
    offsety = 0
    pos.x = pos.x - 1
  }
  if(packet.message=='f'){
    offsetx = 1
    offsety = 0
    pos.x = pos.x + 1
  }
  if(packet.message=='l'){
    offsetx = 0
    offsety = -1
    pos.z = pos.z - 1
  }
  if(packet.message=='r'){
    offsetx = 0
    offsety = 1
    pos.z = pos.z + 1
  }
  if(packet.message=='u'){
    offsetx = 0
    offsety = 1
    pos.y = pos.y + 1
  }
  if(packet.message=='d'){
    offsetx = 0
    offsety = 1
    pos.y = pos.y - 1
  }
  client.queue('player_auth_input', {
  pitch: 0,
  yaw: -60,
  position: pos,
  move_vector: { x: offsetx, z: offsety },
  head_yaw: -70,
  input_data: {
    _value: 0n,
    ascend: false,
    descend: false,
    north_jump: false,
    jump_down: false,
    sprint_down: false,
    change_height: false,
    jumping: false,
    auto_jumping_in_water: false,
    sneaking: false,
    sneak_down: false,
    up: false,
    down: false,
    left: false,
    right: false,
    up_left: false,
    up_right: false,
    want_up: false,
    want_down: false,
    want_down_slow: false,
    want_up_slow: false,
    sprinting: false,
    ascend_block: false,
    descend_block: false,
    sneak_toggle_down: false,
    persist_sneak: false,
    start_sprinting: false,
    stop_sprinting: false,
    start_sneaking: false,
    stop_sneaking: false,
    start_swimming: false,
    stop_swimming: false,
    start_jumping: false,
    start_gliding: false,
    stop_gliding: false,
    item_interact: false,
    block_action: false,
    item_stack_request: false
  },
  input_mode: 'mouse',
  play_mode: 'screen',
  interaction_model: 'crosshair',
  gaze_direction: undefined,
  tick: client.tick,
  delta: { x: offsetx, y: 0, z: offsety },
  transaction: undefined,
  item_stack_request: undefined,
  block_action: undefined
})
})
```

**Reply by frizib** [reply]
yes, relay

**Reply by Jajo006** [reply]
A common example for the relay to also print all movement packages would be
```const { Relay } = require('bedrock-protocol')

// Start the proxy server
const relay = new Relay({
  version: '1.19.40', // The version
  /* host and port to listen for clients on */
  host: 'IP',
  port: 19132,
  /* Where to send upstream packets to */
  destination: {
    host: 'IP',
    port: 19132
  }
})
relay.conLog = console.debug
relay.listen() // Tell the server to start listening.

relay.on('connect', player => {
  console.log('New connection', player.connection.address)

  // Client is sending a message to the server
  player.on('serverbound', ({ name, params }) => {
    if (name == 'player_auth_input') {
      console.log(name);
      console.log(params)
    }
  })
})
```

**Reply by frizib** [reply]
Is there an example of clicking on a block? (For the clicker)

**Reply by Jajo006** [reply]
Hi there,
Sorry for only answering now.

I personally don't know that much because I at the time of experimenting with it myself was happy with my client being able to move.
If you can't find a packet for clicking with the relay, be sure to ask on the PrismarineJS discord server where they have a separate thread for the bedrock protocol. They will most certainly help you out.

I hope this helps.

**Reply by frizib** [reply]
Thanks
